File: tools/verify-objc-api.py
#!/usr/bin/env python3
"""Capture and verify the pinned Objective-C JavaScriptCore header surface."""

from __future__ import annotations

import argparse
import hashlib
import json
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
INVENTORY = ROOT / "docs/objc-api/jsc-objc-api-macos-27.0.json"
FRAMEWORK = Path("System/Library/Frameworks/JavaScriptCore.framework/Headers")
HEADERS = ("JSContext.h", "JSValue.h", "JSVirtualMachine.h", "JSManagedValue.h", "JSExport.h")
VALID_STATUSES = {"implemented", "pending", "platform_gated"}


def fail(message: str) -> None:
    print(f"objc-api audit: {message}", file=sys.stderr)
    raise SystemExit(1)


def sha256(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as stream:
        for chunk in iter(lambda: stream.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def normalized_sha256(path: Path) -> str:
    lines = path.read_bytes().splitlines(keepends=True)
    normalized = b"".join(
        line.rstrip(b" \t\r\n") + (b"\n" if line.endswith((b"\n", b"\r")) else b"")
        for line in lines
    )
    return hashlib.sha256(normalized).hexdigest()


def normalize(text: str) -> str:
    return " ".join(text.split())


def uncomment(text: str) -> str:
    text = re.sub(r"/\*.*?\*/", "", text, flags=re.S)
    return re.sub(r"//[^\n]*", "", text)


def selector(tail: str) -> str:
    parts = re.findall(r"\b([A-Za-z_][A-Za-z0-9_]*)\s*:", tail)
    if parts:
        return ":".join(parts) + ":"
    match = re.search(r"\b([A-Za-z_][A-Za-z0-9_]*)\b", tail)
    if not match:
        fail(f"cannot derive selector from {tail!r}")
    return match.group(1)


def property_name(declaration: str) -> str:
    block = re.search(r"\(\^\s*([A-Za-z_][A-Za-z0-9_]*)\s*\)", declaration)
    if block:
        return block.group(1)
    body = declaration.split(" API_", 1)[0].split(" NS_", 1)[0]
    names = re.findall(r"\b([A-Za-z_][A-Za-z0-9_]*)\b", body)
    if not names:
        fail(f"cannot derive property name from {declaration!r}")
    return names[-1]


def pending_issue(header: str, name: str) -> int:
    if header == "JSManagedValue.h" or name in {
        "addManagedReference:withOwner:",
        "removeManagedReference:withOwner:",
    }:
        return 159
    if header == "JSExport.h":
        return 160
    return 158


def parse_header(header: str, source: str) -> tuple[list[dict[str, object]], list[dict[str, object]]]:
    text = uncomment(source)
    containers: list[dict[str, object]] = []
    declarations: list[dict[str, object]] = []
    block_re = re.compile(
        r"@(interface|protocol)\s+([A-Za-z_][A-Za-z0-9_]*)"
        r"(?:\s*\(([^)]*)\))?(?:\s*:\s*([A-Za-z_][A-Za-z0-9_]*))?"
        r"(?:\s*<([^>]*)>)?(.*?)@end",
        re.S,
    )
    for match in block_re.finditer(text):
        kind, name, category, superclass, adopted, body = match.groups()
        container = name if not category else f"{name}({normalize(category)})"
        prefix = text[max(0, match.start() - 200) : match.start()]
        availability_matches = re.findall(
            r"(?:API_AVAILABLE|NS_CLASS_AVAILABLE)\([^\n]*\)", prefix
        )
        containers.append(
            {
                "header": header,
                "kind": kind,
                "name": name,
                "category": normalize(category) if category else None,
                "superclass": superclass,
                "adopted_protocols": sorted(
                    item.strip() for item in (adopted or "").split(",") if item.strip()
                ),
                "availability": availability_matches[-1] if availability_matches else None,
            }
        )
        for method in re.finditer(r"(?ms)^[ \t]*([+-])\s*\(([^)]*)\)\s*(.*?);", body):
            polarity, return_type, tail = method.groups()
            method_selector = selector(tail)
            declarations.append(
                {
                    "header": header,
                    "container": container,
                    "kind": "class_method" if polarity == "+" else "instance_method",
                    "name": method_selector,
                    "declaration": normalize(f"{polarity} ({return_type}) {tail}"),
                    "status": "pending",
                    "issue": pending_issue(header, method_selector),
                }
            )
        for prop in re.finditer(r"(?ms)^[ \t]*@property\s+(.*?);", body):
            declaration = normalize(f"@property {prop.group(1)}")
            name_value = property_name(declaration)
            declarations.append(
                {
                    "header": header,
                    "container": container,
                    "kind": "property",
                    "name": name_value,
                    "declaration": declaration,
                    "status": "pending",
                    "issue": pending_issue(header, name_value),
                }
            )
    for typedef in re.finditer(r"(?m)^typedef\s+([^;]*\bJSValueProperty)\s*;", text):
        declarations.append(
            {
                "header": header,
                "container": "global",
                "kind": "typedef",
                "name": "JSValueProperty",
                "declaration": normalize(f"typedef {typedef.group(1)}"),
                "status": "pending",
                "issue": 158,
            }
        )
    for symbol in re.finditer(
        r"(?m)^JS_EXPORT\s+extern\s+([^;]*\b(JSPropertyDescriptor[A-Za-z0-9_]+))\s*;",
        text,
    ):
        declarations.append(
            {
                "header": header,
                "container": "global",
                "kind": "data_symbol",
                "name": symbol.group(2),
                "declaration": normalize(f"JS_EXPORT extern {symbol.group(1)}"),
                "status": "pending",
                "issue": 158,
            }
        )
    return containers, declarations


def parse_surface(headers_root: Path) -> tuple[dict[str, str], dict[str, str], list[dict[str, object]], list[dict[str, object]]]:
    containers: list[dict[str, object]] = []
    declarations: list[dict[str, object]] = []
    hashes: dict[str, str] = {}
    normalized_hashes: dict[str, str] = {}
    for header in HEADERS:
        path = headers_root / header
        if not path.is_file():
            fail(f"missing SDK header {path}")
        source = path.read_text()
        parsed_containers, parsed_declarations = parse_header(header, source)
        containers.extend(parsed_containers)
        declarations.extend(parsed_declarations)
        hashes[header] = sha256(path)
        normalized_hashes[header] = normalized_sha256(path)
    if "@protocol JSExport" not in uncomment((headers_root / "JSExport.h").read_text()):
        fail("JSExport protocol is absent")
    declarations.append(
        {
            "header": "JSExport.h",
            "container": "global",
            "kind": "macro",
            "name": "JSExportAs",
            "declaration": "#define JSExportAs(PropertyName, Selector)",
            "status": "pending",
            "issue": 160,
        }
    )
    containers.sort(key=lambda item: (str(item["header"]), str(item["kind"]), str(item["name"]), str(item["category"])))
    declarations.sort(key=lambda item: (str(item["header"]), str(item["container"]), str(item["kind"]), str(item["name"]), str(item["declaration"])))
    return hashes, normalized_hashes, containers, declarations


def capture(sdk_root: Path) -> dict[str, object]:
    hashes, normalized_hashes, containers, declarations = parse_surface(sdk_root / FRAMEWORK)
    return {
        "schema_version": 1,
        "target": {
            "platform": "macOS",
            "sdk_version": "27.0",
            "sdk_build": "26A5368g",
            "captured_at": "2026-07-16",
        },
        "source_headers": hashes,
        "checked_header_hashes": normalized_hashes,
        "containers": containers,
        "declarations": declarations,
    }


def declaration_key(entry: dict[str, object]) -> tuple[str, str, str, str, str]:
    return tuple(str(entry.get(field, "")) for field in ("header", "container", "kind", "name", "declaration"))


def validate(data: dict[str, object]) -> None:
    if data.get("schema_version") != 1:
        fail("schema_version must be 1")
    if set(data.get("source_headers", {})) != set(HEADERS):
        fail("source_headers must name exactly the five targeted Objective-C headers")
    if set(data.get("checked_header_hashes", {})) != set(HEADERS):
        fail("checked_header_hashes must name exactly the five targeted Objective-C headers")
    containers = data.get("containers")
    declarations = data.get("declarations")
    if not isinstance(containers, list) or not isinstance(declarations, list):
        fail("containers and declarations must be arrays")
    container_keys = [
        (entry.get("header"), entry.get("kind"), entry.get("name"), entry.get("category"))
        for entry in containers
    ]
    if len(container_keys) != len(set(container_keys)):
        fail("container entries must be unique")
    declaration_keys = [declaration_key(entry) for entry in declarations]
    if len(declaration_keys) != len(set(declaration_keys)):
        fail("declaration entries must be unique")
    known_containers = {entry.get("name") if not entry.get("category") else f"{entry.get('name')}({entry.get('category')})" for entry in containers}
    for entry in declarations:
        if entry.get("container") != "global" and entry.get("container") not in known_containers:
            fail(f"unknown declaration container {entry.get('container')}")
        status = entry.get("status")
        if status not in VALID_STATUSES:
            fail(f"{entry.get('name')} has invalid status {status!r}")
        if status != "implemented" and not isinstance(entry.get("issue"), int):
            fail(f"unfinished {entry.get('name')} must link a numeric GitHub issue")


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--sdk-root", type=Path, help="SDK root to compare with the pin")
    parser.add_argument("--capture", action="store_true", help="print a fresh inventory from --sdk-root")
    args = parser.parse_args()
    if args.capture:
        if args.sdk_root is None:
            fail("--capture requires --sdk-root")
        print(json.dumps(capture(args.sdk_root), indent=2) + "\n", end="")
        return
    if not INVENTORY.is_file():
        fail(f"missing inventory {INVENTORY.relative_to(ROOT)}")
    data = json.loads(INVENTORY.read_text())
    validate(data)
    _, checked_hashes, checked_containers, checked_declarations = parse_surface(
        ROOT / "include/JavaScriptCore"
    )
    if checked_hashes != data["checked_header_hashes"]:
        fail("checked-in Objective-C header hashes drifted from the pin")
    if checked_containers != data["containers"]:
        fail("checked-in Objective-C containers drifted from the inventory")
    if [declaration_key(entry) for entry in checked_declarations] != [
        declaration_key(entry) for entry in data["declarations"]
    ]:
        fail("checked-in Objective-C declarations drifted from the inventory")
    if args.sdk_root is not None:
        actual = capture(args.sdk_root)
        if actual["source_headers"] != data["source_headers"]:
            fail("pinned Objective-C SDK header hashes drifted")
        if actual["checked_header_hashes"] != data["checked_header_hashes"]:
            fail("normalized Objective-C SDK header hashes drifted")
        if actual["containers"] != data["containers"]:
            fail("pinned Objective-C containers drifted")
        actual_keys = [declaration_key(entry) for entry in actual["declarations"]]
        expected_keys = [declaration_key(entry) for entry in data["declarations"]]
        if actual_keys != expected_keys:
            fail("pinned Objective-C declarations drifted")
    implemented = sum(entry["status"] == "implemented" for entry in data["declarations"])
    pending = len(data["declarations"]) - implemented
    print(
        f"objc-api audit: {len(data['containers'])} containers, "
        f"{len(data['declarations'])} declarations, {implemented} complete, {pending} pending"
    )


if __name__ == "__main__":
    main()

File: tools/verify-objc-api.ts
/** Capture and verify the pinned Objective-C JavaScriptCore header surface. */
import { readText, removeTemporaryDirectory, sha256File, temporaryDirectory, writeText } from "./lib/home";
declare const __dirname: string;
declare const __filename: string;
const ROOT = __dirname === "tools" ? "." : __dirname.slice(0, __dirname.lastIndexOf("/tools"));
const join = (left: string, right: string): string => `${left.replace(/\/$/, "")}/${right}`;
const INVENTORY = join(ROOT, "docs/objc-api/jsc-objc-api-macos-27.0.json"), FRAMEWORK = "System/Library/Frameworks/JavaScriptCore.framework/Headers";
const HEADERS = ["JSContext.h", "JSValue.h", "JSVirtualMachine.h", "JSManagedValue.h", "JSExport.h"], VALID_STATUSES = ["implemented", "pending", "platform_gated"];
const fail = (message: string): never => { throw new Error(`objc-api audit: ${message}`); };
const normalize = (text: string): string => text.trim().split(/\s+/).join(" ");
const uncomment = (text: string): string => text.replace(/\/\*[\s\S]*?\*\//g, "").replace(/\/\/[^\n]*/g, "");
const allMatches = (pattern: RegExp, text: string): RegExpExecArray[] => { const out: RegExpExecArray[] = []; let match: RegExpExecArray | null; while ((match = pattern.exec(text)) !== null) out.push(match); return out; };
const selector = (tail: string): string => { const parts = allMatches(/\b([A-Za-z_][A-Za-z0-9_]*)\s*:/g, tail).map((match) => match[1]); if (parts.length) return parts.join(":") + ":"; const match = /\b([A-Za-z_][A-Za-z0-9_]*)\b/.exec(tail); if (!match) fail(`cannot derive selector from ${JSON.stringify(tail)}`); return match[1]; };
const propertyName = (declaration: string): string => { const block = /\(\^\s*([A-Za-z_][A-Za-z0-9_]*)\s*\)/.exec(declaration); if (block) return block[1]; const body = declaration.split(" API_")[0].split(" NS_")[0], names = allMatches(/\b([A-Za-z_][A-Za-z0-9_]*)\b/g, body).map((match) => match[1]); if (!names.length) fail(`cannot derive property name from ${JSON.stringify(declaration)}`); return names[names.length - 1]; };
const pendingIssue = (header: string, name: string): number => header === "JSManagedValue.h" || ["addManagedReference:withOwner:", "removeManagedReference:withOwner:"].includes(name) ? 159 : header === "JSExport.h" ? 160 : 158;

function parseHeader(header: string, source: string): { containers: any[]; declarations: any[] } {
  const text = uncomment(source), containers: any[] = [], declarations: any[] = [];
  const starts = allMatches(/@(interface|protocol)\s+([A-Za-z_][A-Za-z0-9_]*)/g, text);
  for (const start of starts) {
    const signatureEnd = text.indexOf("\n", start.index), end = text.indexOf("@end", signatureEnd);
    if (end < 0) fail(`${header}: unterminated ${start[1]} ${start[2]}`);
    const signature = text.slice(start.index, signatureEnd), details = /^@(interface|protocol)\s+([A-Za-z_][A-Za-z0-9_]*)(?:\s*\(([^)]*)\))?(?:\s*:\s*([A-Za-z_][A-Za-z0-9_]*))?(?:\s*<([^>]*)>)?/.exec(signature);
    if (!details) fail(`${header}: cannot parse ${signature}`);
    const kind = details[1], name = details[2], category = details[3], superclass = details[4] || null, adopted = details[5], body = text.slice(signatureEnd, end), container = category ? `${name}(${normalize(category)})` : name;
    const prefix = text.slice(Math.max(0, start.index - 200), start.index), availability = allMatches(/(?:API_AVAILABLE|NS_CLASS_AVAILABLE)\([^\n]*\)/g, prefix).map((item) => item[0]);
    containers.push({ header, kind, name, category: category ? normalize(category) : null, superclass, adopted_protocols: (adopted || "").split(",").map((item) => item.trim()).filter(Boolean).sort(), availability: availability.length ? availability[availability.length - 1] : null });
    let pending = "";
    for (const rawLine of body.split("\n")) {
      const line = rawLine.trim();
      if (!pending && (/^[+-]\s*\(/.test(line) || line.startsWith("@property "))) pending = line;
      else if (pending) pending += " " + line;
      if (!pending.includes(";")) continue;
      const declarationText = pending.slice(0, pending.indexOf(";")); pending = "";
      if (declarationText.startsWith("@property ")) {
        const declaration = normalize(declarationText), propertyValue = propertyName(declaration);
        declarations.push({ header, container, kind: "property", name: propertyValue, declaration, status: "pending", issue: pendingIssue(header, propertyValue) });
      } else {
        const method = /^([+-])\s*\(([^)]*)\)\s*([\s\S]*)$/.exec(declarationText);
        if (!method) fail(`${header}: cannot parse method ${declarationText}`);
        const methodSelector = selector(method[3]);
        declarations.push({ header, container, kind: method[1] === "+" ? "class_method" : "instance_method", name: methodSelector, declaration: normalize(`${method[1]} (${method[2]}) ${method[3]}`), status: "pending", issue: pendingIssue(header, methodSelector) });
      }
    }
  }
  for (const typedef of allMatches(/^typedef\s+([^;]*\bJSValueProperty)\s*;/gm, text)) declarations.push({ header, container: "global", kind: "typedef", name: "JSValueProperty", declaration: normalize(`typedef ${typedef[1]}`), status: "pending", issue: 158 });
  for (const symbol of allMatches(/^JS_EXPORT\s+extern\s+([^;]*\b(JSPropertyDescriptor[A-Za-z0-9_]+))\s*;/gm, text)) declarations.push({ header, container: "global", kind: "data_symbol", name: symbol[2], declaration: normalize(`JS_EXPORT extern ${symbol[1]}`), status: "pending", issue: 158 });
  return { containers, declarations };
}

function normalizedSha(file: string): string {
  const text = readText(file), lines = text.match(/[^\n]*\n|[^\n]+$/g) || [], normalized = lines.map((line) => line.replace(/[ \t\r\n]+$/, "") + (line.endsWith("\n") || line.endsWith("\r") ? "\n" : "")).join("");
  const directory = temporaryDirectory("zig-js-objc-hash");
  try { const output = join(directory, "normalized.h"); writeText(output, normalized); return sha256File(output); } finally { removeTemporaryDirectory(directory); }
}
const declarationKey = (entry: any): string[] => ["header", "container", "kind", "name", "declaration"].map((field) => String(entry[field] || ""));
const compareKeys = (left: any[], right: any[]): number => { const a = left.map(String).join("\u0000"), b = right.map(String).join("\u0000"); return a < b ? -1 : a > b ? 1 : 0; };
function parseSurface(root: string): any {
  const hashes: Record<string, string> = {}, normalized: Record<string, string> = {}, containers: any[] = [], declarations: any[] = [];
  for (const header of HEADERS) {
    const file = join(root, header), parsed = parseHeader(header, readText(file));
    containers.push(...parsed.containers); declarations.push(...parsed.declarations); hashes[header] = sha256File(file); normalized[header] = normalizedSha(file);
  }
  if (!uncomment(readText(join(root, "JSExport.h"))).includes("@protocol JSExport")) fail("JSExport protocol is absent");
  declarations.push({ header: "JSExport.h", container: "global", kind: "macro", name: "JSExportAs", declaration: "#define JSExportAs(PropertyName, Selector)", status: "pending", issue: 160 });
  containers.sort((a, b) => compareKeys([a.header, a.kind, a.name, a.category === null ? "None" : a.category], [b.header, b.kind, b.name, b.category === null ? "None" : b.category]));
  declarations.sort((a, b) => compareKeys(declarationKey(a), declarationKey(b)));
  return { hashes, normalized, containers, declarations };
}
function capture(sdkRoot: string): any { const surface = parseSurface(join(sdkRoot, FRAMEWORK)); return { schema_version: 1, target: { platform: "macOS", sdk_version: "27.0", sdk_build: "26A5368g", captured_at: "2026-07-16" }, source_headers: surface.hashes, checked_header_hashes: surface.normalized, containers: surface.containers, declarations: surface.declarations }; }
function validate(data: any): void {
  if (data.schema_version !== 1) fail("schema_version must be 1");
  if (JSON.stringify(Object.keys(data.source_headers || {}).sort()) !== JSON.stringify(HEADERS.slice().sort())) fail("source_headers must name exactly the five targeted Objective-C headers");
  if (JSON.stringify(Object.keys(data.checked_header_hashes || {}).sort()) !== JSON.stringify(HEADERS.slice().sort())) fail("checked_header_hashes must name exactly the five targeted Objective-C headers");
  if (!Array.isArray(data.containers) || !Array.isArray(data.declarations)) fail("containers and declarations must be arrays");
  const containerKeys = data.containers.map((entry: any) => JSON.stringify([entry.header, entry.kind, entry.name, entry.category])); if (new Set(containerKeys).size !== containerKeys.length) fail("container entries must be unique");
  const declarationKeys = data.declarations.map((entry: any) => JSON.stringify(declarationKey(entry))); if (new Set(declarationKeys).size !== declarationKeys.length) fail("declaration entries must be unique");
  const known = data.containers.map((entry: any) => entry.category ? `${entry.name}(${entry.category})` : entry.name);
  for (const entry of data.declarations) { if (entry.container !== "global" && !known.includes(entry.container)) fail(`unknown declaration container ${entry.container}`); if (!VALID_STATUSES.includes(entry.status)) fail(`${entry.name} has invalid status ${JSON.stringify(entry.status)}`); if (entry.status !== "implemented" && !Number.isInteger(entry.issue)) fail(`unfinished ${entry.name} must link a numeric GitHub issue`); }
}

export function verifyObjCAPI(sdkRoot = "", captureOnly = false): void {
  if (captureOnly) { if (!sdkRoot) fail("--capture requires --sdk-root"); process.stdout.write(JSON.stringify(capture(sdkRoot), null, 2) + "\n"); return; }
  const data = JSON.parse(readText(INVENTORY)); validate(data);
  const checked = parseSurface(join(ROOT, "include/JavaScriptCore"));
  if (JSON.stringify(checked.normalized) !== JSON.stringify(data.checked_header_hashes)) fail("checked-in Objective-C header hashes drifted from the pin");
  if (JSON.stringify(checked.containers) !== JSON.stringify(data.containers)) fail("checked-in Objective-C containers drifted from the inventory");
  if (JSON.stringify(checked.declarations.map(declarationKey)) !== JSON.stringify(data.declarations.map(declarationKey))) fail("checked-in Objective-C declarations drifted from the inventory");
  if (sdkRoot) { const actual = capture(sdkRoot); if (JSON.stringify(actual.source_headers) !== JSON.stringify(data.source_headers)) fail("pinned Objective-C SDK header hashes drifted"); if (JSON.stringify(actual.checked_header_hashes) !== JSON.stringify(data.checked_header_hashes)) fail("normalized Objective-C SDK header hashes drifted"); if (JSON.stringify(actual.containers) !== JSON.stringify(data.containers)) fail("pinned Objective-C containers drifted"); if (JSON.stringify(actual.declarations.map(declarationKey)) !== JSON.stringify(data.declarations.map(declarationKey))) fail("pinned Objective-C declarations drifted"); }
  const implemented = data.declarations.filter((entry: any) => entry.status === "implemented").length;
  console.log(`objc-api audit: ${data.containers.length} containers, ${data.declarations.length} declarations, ${implemented} complete, ${data.declarations.length - implemented} pending`);
}
if (process.argv[1] === __filename) {
  let sdkRoot = "", captureOnly = false; const args = process.argv.slice(2);
  for (let index = 0; index < args.length; index += 1) { if (args[index] === "--sdk-root" && args[index + 1]) sdkRoot = args[++index]; else if (args[index] === "--capture") captureOnly = true; else fail(`unknown argument: ${args[index]}`); }
  verifyObjCAPI(sdkRoot, captureOnly);
}
